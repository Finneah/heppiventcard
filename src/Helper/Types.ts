export type StampCardType = {
  id?: string;
  title: string;
  complete: Boolean;
  date_of_creation: Date;
  date_of_completed?: Date;
  completed_image?: string;
};

export interface SectionPart extends StampCardType {
  content: StampType[] | null | undefined;
}

export type StampType = {
  id?: string;
  number: number;
  name: string | null | undefined;
  image: string | null | undefined;
  picture: string | null | undefined;
  done: Boolean;
  date: Date | undefined;
  description: string | null | undefined;
  stampCard: StampCardType;
};

export type UserType = {
  id?: number;
  name: string | null | undefined;
  rank: number;
};

export type QRCodeObject = {
  url: string;
  date: string;
  description: string;
  name: string;
  picture: string | null | undefined;
};
export type QRCodeReturnData = {
  image: string;
  description: string;
  date: Date;
  done: boolean;
  name: string;
};
