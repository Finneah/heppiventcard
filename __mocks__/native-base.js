jest.mock('native-base', () => {
  return {
    StyleProvider: jest.fn().mockImplementation((x) => x),
  };
});
