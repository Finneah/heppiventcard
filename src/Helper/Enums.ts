export enum DEBUGLEVEL {
  OFF,
  DATA,
  ALL,
}
export enum UserRank {
  RANK_00 = 0,
  RANK_03 = 3,
  RANK_10 = 10,
  RANK_15 = 15,
  RANK_20 = 20,
  RANK_30 = 30,
}
