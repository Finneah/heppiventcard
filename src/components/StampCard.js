import {
  Body,
  Button,
  Card,
  CardItem,
  Content,
  Icon,
  Left,
  Text,
  Thumbnail,
  Title,
} from 'native-base';
import React, {useState} from 'react';
import {
  Dimensions,
  Image,
  Pressable,
  StyleSheet,
  TouchableWithoutFeedback,
  View,
} from 'react-native';
import GlobalColors from '../styles/GlobalColors';
import bg from '../image/Download.jpeg';
import Modal from 'react-native-modal-patch';
import QRCodeScanner from 'react-native-qrcode-scanner';
import {RNCamera} from 'react-native-camera';
import {strings} from '../i18n';
import RNFetchBlob from 'rn-fetch-blob';
import {StampCards, Stamps, User} from '../database';
const numCol = 3;

const StampCard = ({item}) => {
  const [stampCard, setStampCard] = useState(item);
  const [selectedItem, setSelectedItem] = useState(undefined);
  const [modalVisible, setModalVisible] = useState(false);
  const [qrCodeModalVisible, setQRCodeModalVisible] = useState(false);

  var doneItems = 0;
  if (item.content) {
    item.content.forEach((content) => {
      if (content.done !== 0) {
        doneItems++;
      }
    });
  }

  function _showDetails(item, index) {
    if (selectedItem === index) {
      setSelectedItem(undefined);
    } else {
      console.log('_showDetails', item);
      setSelectedItem(item);
    }

    if (item.done) {
      setModalVisible(true);
    }
  }

  function _onSuccessReadQRCode(e) {
    var data = JSON.parse(e.data);
    console.log(data);
    RNFetchBlob.fetch('GET', data.url, {})
      .then((res) => {
        let status = res.info().status;

        if (status === 200) {
          selectedItem.image = res.data;
          selectedItem.description = data.description;
          selectedItem.date = new Date(data.date);
          selectedItem.done = 1;
          selectedItem.name = data.name;
          Stamps.update(selectedItem, true);
        } else {
          // handle other status codes
        }
      })

      // Something went wrong:
      .catch((errorMessage, statusCode) => {
        // error handling
        console.info(errorMessage, statusCode);
      });

    // var example = {
    //   url:
    //     'https://previews.123rf.com/images/chrisdorney/chrisdorney1607/chrisdorney16â€¦1360474-abgeschlossen-stempel-%C3%BCber-einen-wei%C3%9Fen-hintergrund-.jpg',
    //   date: '2020-12-10',
    //   description: 'Textblabla',
    // };

    var x = JSON.parse(e.data);

    setQRCodeModalVisible(false);
  }

  return (
    <View style={styles.content}>
      {item.content
        ? item.content.map((i, index) => (
            <>
              <Pressable
                key={(i.number + index).toString()}
                style={[styles.item]}
                onPress={() => {
                  console.log(i, index);
                  _showDetails(i, index);
                }}>
                {i.done !== 0 ? (
                  <Image
                    key={(i.number + index).toString()}
                    source={{uri: 'data:image/png;base64,' + i.image}}
                    style={styles.image}
                  />
                ) : (
                  <View
                    key={(i.number + index).toString()}
                    style={[styles.image, styles.stampItem]}>
                    {doneItems === index ? (
                      <Icon
                        onPress={() => {
                          setQRCodeModalVisible(true);
                          setSelectedItem(i);
                        }}
                        style={{color: GlobalColors.brandPrimary}}
                        name="add"
                      />
                    ) : (
                      <Title style={{color: GlobalColors.brandPrimary}}>
                        {i.number.toString()}
                      </Title>
                    )}
                  </View>
                )}
              </Pressable>
            </>
          ))
        : null}
      {item.finished ? (
        <View style={[styles.item, styles.lastItem]}>
          <Image source={item.finishedIcon} style={styles.image} />
        </View>
      ) : (
        <View style={[styles.item, styles.lastItem]}>
          <Text>{strings('FINISH_CARD_TEXT')}</Text>
        </View>
      )}

      <Modal
        transparent
        animationType="slide"
        presentationStyle="overFullScreen" // <-- Swipe down/dismiss works now!
        visible={modalVisible}
        onDismiss={() => setModalVisible(false)} // <-- This gets called all the time
        onRequestClose={() => setModalVisible(false)}>
        <TouchableWithoutFeedback onPress={() => setModalVisible(false)}>
          <View style={styles.centeredView}>
            <View style={styles.modalView}>
              <Card transparent>
                <CardItem header first>
                  <Left>
                    {selectedItem ? (
                      <Thumbnail
                        large
                        source={{
                          uri: 'data:image/png;base64,' + selectedItem.image,
                        }}
                      />
                    ) : null}
                    <Body>
                      <Title style={{color: GlobalColors.dark}}>
                        {selectedItem ? selectedItem.name : ''}
                      </Title>
                      <Text note>
                        {selectedItem
                          ? selectedItem.date.toLocaleDateString()
                          : ''}
                      </Text>
                    </Body>
                  </Left>
                </CardItem>
                <CardItem cardBody>
                  <Text note>
                    {selectedItem ? selectedItem.description : ''}
                  </Text>
                </CardItem>
                <CardItem footer last>
                  <Button
                    style={{marginTop: 10}}
                    centered
                    rounded
                    onPress={() => {
                      setModalVisible(false);
                    }}>
                    <Text>{strings('CLOSE')}</Text>
                  </Button>
                </CardItem>
              </Card>
            </View>
          </View>
        </TouchableWithoutFeedback>
      </Modal>

      <Modal
        animationType="slide"
        presentationStyle="overFullScreen" // <-- Swipe down/dismiss works now!
        visible={qrCodeModalVisible}
        onDismiss={() => setQRCodeModalVisible(false)} // <-- This gets called all the time
        onRequestClose={() => setQRCodeModalVisible(false)}>
        <View style={{padding: 20, flex: 1, flexDirection: 'column'}}>
          <Card>
            <CardItem first last>
              <Left>
                <Body>
                  <Text>{strings('SCAN_QR_CODE_NOW')}</Text>
                  <Text note>{strings('QR_CODE_DESCRIPTION')}</Text>
                </Body>
              </Left>
            </CardItem>
          </Card>

          <QRCodeScanner
            // cameraStyle={{
            //   alignSelf: 'center',
            // }}
            // containerStyle={{flex: 1}}
            fadeIn={true}
            onRead={_onSuccessReadQRCode.bind(this)}
            flashMode={RNCamera.Constants.FlashMode.torch}
            topContent={null}
            bottomContent={null}
          />
          <Button
            centered
            rounded
            onPress={() => {
              setQRCodeModalVisible(false);
            }}>
            <Text>{strings('CANCEL')}</Text>
          </Button>
        </View>
      </Modal>
    </View>
  );
};

const styles = StyleSheet.create({
  centerText: {
    flex: 1,
    color: '#777',
    alignSelf: 'center',
  },
  detail: {
    position: 'absolute',
    top: 0,
    borderRadius: 20,
    zIndex: 1,
    backgroundColor: 'rgba(248,248,248,0.3)',
    width: Dimensions.get('screen').width * ((100 / numCol - 10) / 100) * 3.5,
    height: Dimensions.get('screen').width * ((100 / numCol - 10) / 100) * 2.1,
  },
  item: {
    zIndex: 0,
    flexDirection: 'column',
    backgroundColor: 'transparent',
    margin: 5,

    padding: 0,
    borderRadius: 20,
    width: Dimensions.get('screen').width * ((100 / numCol - 10) / 100),
    height: Dimensions.get('screen').width * ((100 / numCol - 10) / 100),
  },
  content: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    alignContent: 'space-between',
    justifyContent: 'space-evenly',
  },
  lastItem: {
    backgroundColor: GlobalColors.lightGrey,
    alignItems: 'center',
    justifyContent: 'center',
    width: Dimensions.get('screen').width * ((100 / numCol - 10) / 100) * 2.2,
  },
  image: {
    backgroundColor: GlobalColors.lightGrey,
    borderRadius: 20,
    width: '100%',
    height: '100%',
    justifyContent: 'center',
  },
  stampItem: {
    borderRadius: 20,
    justifyContent: 'center',
    alignItems: 'center',
  },
  transparentBorder: {borderColor: 'transparent'},
  transparentBG: {backgroundColor: 'transparent'},
  centeredView: {
    flex: 1,
    justifyContent: 'flex-end',
    alignItems: 'center',
    paddingTop: 22,
    backgroundColor: GlobalColors.dark_opacity,
  },
  modalView: {
    minHeight: Dimensions.get('screen').height * 0.4,
    width: Dimensions.get('screen').width,
    backgroundColor: 'white',
    borderTopLeftRadius: 20,
    borderTopRightRadius: 20,
    padding: 20,
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 0.25,
    shadowRadius: 3.84,
    elevation: 5,
  },
});
export default StampCard;
