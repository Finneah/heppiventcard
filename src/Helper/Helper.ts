import {StampCards, Stamps} from '../database';
import {UserRank} from './Enums';
import {strings} from '../locale/i18n';
import {SectionPart, StampCardType, UserType} from './Types';
import moment from 'moment';
import 'moment/locale/de';
export const isValidUser = (user: UserType): Boolean => {
  try {
    if (user?.name !== '' || user?.rank >= 0) {
      return true;
    }
    throw new Error('name or rank is missing');
  } catch (error) {
    return false;
  }
};
/**
 * "RANK_0": "Gast",
 * "RANK_3": "Stammgast",
 * "RANK_10": "Mitglied",
 * "RANK_15": "Stammmitglied",
 * "RANK_20": "Platinmitglied",
 * "RANK_30": "Goldmitglied",
 */
export const _getRank = (user: UserType): string => {
  let returnValue = strings('RANK_00');
  if (user?.rank) {
    if (user.rank >= UserRank.RANK_30) {
      returnValue = strings('RANK_30');
    } else if (user.rank >= UserRank.RANK_20) {
      returnValue = strings('RANK_20');
    } else if (user.rank >= UserRank.RANK_15) {
      returnValue = strings('RANK_15');
    } else if (user.rank >= UserRank.RANK_10) {
      returnValue = strings('RANK_10');
    } else if (user.rank >= UserRank.RANK_03) {
      returnValue = strings('RANK_03');
    } else {
      returnValue = strings('RANK_00');
    }
  }
  return returnValue;
};

export const _getImageNameForRank = (user: UserType): string => {
  if (user?.rank >= 30) {
    return 'Gold';
  } else if (user.rank >= 20) {
    return 'Platin';
  }
  return 'Standard';
};

export const _checkDoneItems = (stampCard: SectionPart): number | string => {
  try {
    var done = 0;

    if (stampCard?.content) {
      stampCard?.content?.forEach((content: any) => {
        if (content.done === 1 || content.done === true) {
          done++;
        }
      });

      return done;
    } else {
      throw new Error('Error on _checkDoneItems');
    }
  } catch (error) {
    console.warn('_checkDoneItems', error);
    return '_checkDoneItems' + ' ' + error.message;
  }
};

export const _getStampCardTitle = (stampCard: SectionPart): string => {
  let firstDate: Date | undefined;
  let lastDate: Date | undefined;
  let title = '';

  if (stampCard?.content) {
    for (const element of stampCard.content) {
      if (!firstDate && element.date) {
        firstDate = element.date;
      } else {
        if (
          firstDate &&
          element?.date !== null &&
          element?.date !== undefined &&
          element?.date < firstDate
        ) {
          firstDate = element.date;
        }
      }
      if (!lastDate) {
        lastDate = element.date;
      } else {
        if (
          lastDate &&
          element?.date !== null &&
          element?.date !== undefined &&
          element?.date > lastDate
        ) {
          lastDate = element.date;
        }
      }

      if (firstDate && lastDate) {
        title =
          moment(firstDate).locale('de').format('L') +
          ' - ' +
          moment(lastDate).locale('de').format('L');
        console.log(title);
      }
    }
  } else {
    return title;
  }

  return title;
};

export const _createNewStampCard = async () => {
  var newStampCard: StampCardType = {
    title: '',
    complete: false,
    date_of_creation: new Date(),
    date_of_completed: undefined,
    completed_image: undefined,
  };

  newStampCard.date_of_creation = new Date();
  newStampCard.title = '';
  newStampCard.complete = false;

  var card = await StampCards.insert(newStampCard, true)[0];

  var newStamps = [];
  for (let i = 1; i <= 10; i++) {
    newStamps.push({
      number: i,
      done: 0,
      stampCard: card,
    });
  }

  await Stamps.insert(newStamps, true);
};
