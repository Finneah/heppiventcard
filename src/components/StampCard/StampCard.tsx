import * as React from 'react';
import {Dimensions, View, StyleSheet} from 'react-native';
import {Box, Image, Text} from 'native-base';

import {strings} from '../../locale/i18n';

import {SectionPart, StampType} from '../../Helper/Types';

import GlobalColors from '../../styles/GlobalColors';

import {StampItem} from '../StampItem/StampItem';

import SpinnerOverlay from '../Spinner/SpinnerOverlay';
import {_getStampCardTitle} from '../../Helper/Helper';

const numCol = 3;

type Props = {
  item: SectionPart;
};
/**
 * @category Component
 * @namespace StampCard
 * @description StampCard Element
 */
const StampCard: React.FC<Props> = ({item}) => {
  let stampCard = item;
  let stampCardImage = require('../../image/Standard.png');
  const [showLoadingSpinner, setShowLoadingSpinner] = React.useState(false);
  if (stampCard.completed_image) {
    switch (stampCard.completed_image) {
      case 'Standard':
        stampCardImage = require('../../image/Standard.png');
        break;
      case 'Platin':
        stampCardImage = require('../../image/Platin.png');
        break;
      case 'Gold':
        stampCardImage = require('../../image/Gold.png');
        break;

      default:
        break;
    }
  }

  // React.useEffect(() => {
  //   stampCard.title = _getStampCardTitle(stampCard);

  //   // eslint-disable-next-line react-hooks/exhaustive-deps
  // }, [stampCard]);

  /**
   * @memberof StampCard
   */
  function StampCardItems(params: any) {
    let {sCard} = params;
    var stampCardItems = null;

    if (sCard?.content) {
      stampCardItems = sCard.content.map((stamp: StampType, index: number) => (
        <StampItem stamp={stamp} index={index} stampCard={stampCard} />
      ));
    }

    return stampCardItems;
  }

  return (
    <>
      <View style={styles.content}>
        <StampCardItems sCard={stampCard} />
        {stampCard.complete && stampCardImage ? (
          <Box style={[styles.item, styles.lastItem]}>
            <Image
              alt="test"
              resizeMode="stretch"
              style={[
                styles.image,
                {
                  width:
                    Dimensions.get('screen').width *
                    ((100 / numCol - 10) / 100) *
                    2.2,
                  height:
                    Dimensions.get('screen').width *
                    ((100 / numCol - 10) / 100),
                },
              ]}
              source={stampCardImage}
            />
          </Box>
        ) : (
          <Box p={5} style={[styles.item, styles.lastItem]}>
            <Text italic color="secondary.400" textAlign="center">
              {strings('FINISH_CARD_TEXT')}
            </Text>
          </Box>
        )}
      </View>
      <SpinnerOverlay
        spinnerOverlayIsOpen={showLoadingSpinner}
        setSpinnerOverlayIsOpen={setShowLoadingSpinner}
      />
    </>
  );
};
const styles = StyleSheet.create({
  item: {
    zIndex: 0,
    flexDirection: 'column',
    backgroundColor: GlobalColors.lightGrey,
    margin: 5,
    padding: 0,
    borderRadius: 20,
    width: Dimensions.get('screen').width * ((100 / numCol - 10) / 100),
    height: Dimensions.get('screen').width * ((100 / numCol - 10) / 100),
  },
  content: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    alignContent: 'space-between',
    justifyContent: 'space-evenly',
  },
  lastItem: {
    backgroundColor: GlobalColors.lightGrey,
    alignItems: 'center',
    justifyContent: 'center',
    width: Dimensions.get('screen').width * ((100 / numCol - 10) / 100) * 2.2,
  },
  image: {
    backgroundColor: GlobalColors.lightGrey,
    borderRadius: 20,
    width: '100%',
    height: '100%',
  },
  stampItem: {
    borderRadius: 20,
    justifyContent: 'center',
    alignItems: 'center',
  },
  transparentBorder: {borderColor: 'transparent'},
  transparentBG: {backgroundColor: 'transparent'},
  centeredView: {
    flex: 1,
    justifyContent: 'flex-end',
    alignItems: 'center',
    paddingTop: 22,
    backgroundColor: GlobalColors.dark_opacity,
  },
  modalView: {
    minHeight: Dimensions.get('screen').height * 0.4,
    width: Dimensions.get('screen').width,
    backgroundColor: 'white',
    borderTopLeftRadius: 20,
    borderTopRightRadius: 20,
    padding: 20,
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 0.25,
    shadowRadius: 3.84,
    elevation: 5,
  },
});
export default StampCard;
