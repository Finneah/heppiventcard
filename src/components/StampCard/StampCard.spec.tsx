/**
 * @format
 */

import 'react-native';
import * as React from 'react';

import {render} from '@testing-library/react-native';

import {SectionPart, StampCardType, StampType} from '../../Helper/Types';
import {NativeBaseProvider} from 'native-base';
import StampCard from './StampCard';

const wrapper = ({children}: any) => (
  <NativeBaseProvider
    initialWindowMetrics={{
      frame: {x: 0, y: 0, width: 0, height: 0},
      insets: {top: 0, left: 0, right: 0, bottom: 0},
    }}>
    {children}
  </NativeBaseProvider>
);
describe('StampCard', () => {
  test('renders correctly', () => {
    let stampCard: StampCardType = {
      title: 'string',
      complete: false,
      date_of_creation: new Date(),
      date_of_completed: undefined,
      completed_image: undefined,
    };

    let stamp: StampType = {
      number: 1,
      name: undefined,
      image: undefined,
      done: false,
      picture: undefined,
      date: new Date(),
      description: 'string',
      stampCard: stampCard,
    };

    let content: StampType[] = [stamp];
    let item: SectionPart = {
      id: 'string',
      title: 'string',
      complete: false,
      date_of_creation: new Date(),
      date_of_completed: undefined,
      completed_image: 'string',
      content: content,
    };

    const stampCardTest = render(<StampCard item={item} />, {wrapper});
    expect(stampCardTest.toJSON()).toMatchSnapshot();
  });
});
