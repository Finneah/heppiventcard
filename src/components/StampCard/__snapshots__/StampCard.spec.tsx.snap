// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QRCodeActionSheet renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "alignContent": "space-between",
        "flexDirection": "row",
        "flexWrap": "wrap",
        "justifyContent": "space-evenly",
      }
    }
  >
    <View>
      <View
        style={
          Array [
            Object {},
            Object {
              "backgroundColor": "#F8F8F8",
              "borderRadius": 20,
              "flexDirection": "column",
              "height": 175.00000000000003,
              "margin": 5,
              "padding": 0,
              "width": 175.00000000000003,
              "zIndex": 0,
            },
            Object {
              "alignItems": "center",
              "borderRadius": 20,
              "justifyContent": "center",
            },
          ]
        }
      >
        <Text
          allowFontScaling={false}
          fontSize="32px"
          lineHeight="32px"
          onPress={[Function]}
          style={
            Array [
              Object {
                "color": "primary.600",
                "fontSize": 12,
              },
              Array [
                Object {
                  "color": "#164e63",
                  "fontSize": 32,
                  "height": 32,
                  "lineHeight": 32,
                  "width": 32,
                },
              ],
              Object {
                "fontFamily": "Ionicons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          ï„€
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {},
          Object {
            "backgroundColor": "#F8F8F8",
            "borderRadius": 20,
            "flexDirection": "column",
            "height": 175.00000000000003,
            "margin": 5,
            "padding": 0,
            "width": 175.00000000000003,
            "zIndex": 0,
          },
          Object {
            "alignItems": "center",
            "backgroundColor": "#F8F8F8",
            "justifyContent": "center",
            "width": 385.0000000000001,
          },
          Object {
            "padding": 5,
          },
        ]
      }
    >
      <Text
        color="muted.700"
        fontFamily="body"
        fontSize="md"
        fontStyle="normal"
        fontWeight={400}
        style={
          Array [
            Object {
              "color": "#404040",
              "fontSize": 16,
              "fontStyle": "normal",
              "fontWeight": "400",
            },
            Object {
              "textAlign": "center",
            },
          ]
        }
      >
        FINISH_CARD_TEXT
      </Text>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
